test/boost/cdc_generation_test.cc

```diff
@@ -0,0 +1,165 @@
+
+#include "cdc/generation.hh"
+#include "test/lib/random_utils.hh"
+
+namespace cdc {
+
+size_t limit_of_streams_in_topology_description();
+topology_description limit_number_of_streams_if_needed(topology_description&& desc);
+
+} // namespace cdc
+
+static cdc::topology_description create_description(const std::vector<size_t>& streams_count_per_vnode) {
+    std::vector<cdc::token_range_description> result;
+    result.reserve(streams_count_per_vnode.size());
+    size_t vnode_index = 0;
+    int64_t token = std::numeric_limits<int64_t>::min() + 100;
+    for (size_t streams_count : streams_count_per_vnode) {
+        std::vector<cdc::stream_id> streams(streams_count);
+        token += 500;
+        for (size_t idx = 0; idx < streams_count; ++idx) {
+            streams[idx] = cdc::stream_id{dht::token::from_int64(token), vnode_index};
+            token += 100;
+        }
+        token += 10000;
+        // sharding_ignore_msb should matter for limit_number_of_streams_if_needed
+        // so we're using sharding_ignore_msb equal to 0.
+        result.push_back(
+                cdc::token_range_description{dht::token::from_int64(token), std::move(streams), uint8_t{0}});
+        ++vnode_index;
+    }
+    return cdc::topology_description(std::move(result));
+}
+
+static void assert_streams_count(const cdc::topology_description& desc, const std::vector<size_t>& expected_count) {
+    BOOST_REQUIRE_EQUAL(expected_count.size(), desc.entries().size());
+
+    for (size_t idx = 0; idx < expected_count.size(); ++idx) {
+        BOOST_REQUIRE_EQUAL(expected_count[idx], desc.entries()[idx].streams.size());
+    }
+}
+
+static void assert_stream_ids_in_right_token_ranges(const cdc::topology_description& desc) {
+    dht::token start = desc.entries().back().token_range_end;
+    dht::token end = desc.entries().front().token_range_end;
+    for (auto& stream : desc.entries().front().streams) {
+        dht::token t = stream.token();
+        if (t > end) {
+            BOOST_REQUIRE(start < t);
+        } else {
+            BOOST_REQUIRE(t <= end);
+        }
+    }
+    for (size_t idx = 1; idx < desc.entries().size(); ++idx) {
+        for (auto& stream : desc.entries()[idx].streams) {
+            BOOST_REQUIRE(desc.entries()[idx - 1].token_range_end < stream.token());
+            BOOST_REQUIRE(stream.token() <= desc.entries()[idx].token_range_end);
+        }
+    }
+
+}
+
+cdc::stream_id get_stream(const std::vector<cdc::token_range_description>& entries, dht::token tok);
+
+static void assert_random_tokens_mapped_to_streams_with_tokens_in_the_same_token_range(const cdc::topology_description& desc) {
+    for (size_t count = 0; count < 100; ++count) {
+        int64_t token_value = tests::random::get_int(std::numeric_limits<int64_t>::min(), std::numeric_limits<int64_t>::max());
+        dht::token t = dht::token::from_int64(token_value);
+        auto stream = get_stream(desc.entries(), t);
+        auto& e = desc.entries().at(stream.index());
```

Comment:
a stream ID's index is just an optimization for the clients, I don't think we should use it to check invariants. Well we can, but it's optional and not so important.
The more important thing is to check the stream ID's **token**. This test should at least check that the stream ID's token and the random token `t` are in the same token range.

Reply:
It is already checked by each test that each stream belongs to the token range that has this stream in its streams field so checking that token belongs to that token range too means both stream and token belong to the same token range, no?

Reply:
> checking that token belongs to that token range too

to **which** token range? This test uses the token range ending at `desc.entries()[stream.index()].token_range_end`, but nothing here checks that `desc.entries()[stream.index()].streams` contains `stream`.

if we assume that `stream.index()` satisfies the property that `stream` is inside the vector `desc.entries()[stream.index()].streams`, then yes, what you said follows from `assert_stream_ids_in_right_token_ranges`.
But nothing actually checks that property. I don't know if we should assume it. It's not obvious.

Reply:
> But nothing actually checks that property. I don't know if we should assume it. It's not obvious.

I thought that's the definition of index() but I can check it - no harm in that.

Reply:
Also it's not true what you say. I'm already checking that `stream` is inside `desc.entries()[stream.index()].streams` -> `BOOST_REQUIRE(std::find(e.streams.begin(), e.streams.end(), stream) != e.streams.end());`

Reply:
Well storing the index as part of the stream ID format is a client-side performance optimization and I'm not sure if we should depend on it as a correctness invariant. If we do, then a test should guard it.

Reply:
> BOOST_REQUIRE(std::find(e.streams.begin(), e.streams.end(), stream) != e.streams.end())

Again you're implicitly assuming that `e = desc.entries()[stream.index]`.

Reply:
> Well storing the index as part of the stream ID format is a client-side performance optimization and I'm not sure if we should depend on it as a correctness invariant. If we do, then a test should guard it.

Look one line below.

Reply:
Ah nevermind, you're talking about this test now.
Yes, you're right :P

Reply:
> > BOOST_REQUIRE(std::find(e.streams.begin(), e.streams.end(), stream) != e.streams.end())
> 
> Again you're implicitly assuming that `e = desc.entries()[stream.index()]`.

No. I'm setting e to `desc.entries()[stream.index()]` and then verify that this entry contains stream.
Check that each stream in `desc.entries()[stream.index()]` has token that is in token range reflected by desc.entry that contains this stream is already checked in `assert_stream_ids_in_right_token_ranges `.

What else is not verified in your opinion?

Reply:
Ok it looks like everything's covered:
1. `t` is mapped to `stream`
2. `e` contains `stream` (by `BOOST_REQUIRE(std::find(e.streams.begin(), e.streams.end(), stream) != e.streams.end());`)
3. each entry in `e.streams` is in the token range ending at `e.token_range_end` (by the other assert)
4. by 2 and 3, `stream` is in the token range ending at `e.token_range_end`
5. `t` is in the token range ending at `e.token_range_end` by `BOOST_REQUIRE(t <= e.token_range_end);` (except for the special case of `stream.index() = 0`)
6. therefore `t` and `stream` are in the same token range

requires a bit complex reasoning, but works